File: desktop/macos/Sources/NexusDesktop/Diagnostics/DesktopDiagnosticsReport.swift
import Foundation

enum DesktopDiagnosticsReport {
  static func make(
    runtime: SidecarRuntimeConfig?,
    reason: String? = nil,
    startupTimeline: DesktopStartupTimeline? = nil,
    details: [String: Any] = [:]
  ) throws -> String {
    var payload: [String: Any] = [
      "generated_at": ISO8601DateFormatter().string(from: Date()),
      "app": appPayload(),
      "process": processPayload(),
      "system": systemPayload(),
      "paths": pathsPayload(),
      "checks": checksPayload(),
    ]
    if let runtime {
      payload["runtime"] = runtimePayload(runtime)
    }
    if let reason, !reason.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      payload["reason"] = reason
    }
    if let startupTimeline {
      payload["startup_timeline"] = startupTimeline.snapshot()
    }
    if !details.isEmpty {
      payload["details"] = details
    }
    return try jsonString(payload)
  }

  static func writeStartupFailure(error: Error, startupTimeline: DesktopStartupTimeline? = nil) -> URL? {
    do {
      let directory = logsDirectory()
      try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
      let fileURL = directory.appendingPathComponent("startup-failure-\(timestampString()).json")
      let text = try make(runtime: nil, reason: error.localizedDescription, startupTimeline: startupTimeline)
      try text.write(to: fileURL, atomically: true, encoding: .utf8)
      return fileURL
    } catch {
      NSLog("[Nexus Diagnostics] failed to write startup diagnostics: \(error.localizedDescription)")
      return nil
    }
  }

  static func writeRuntimeIssue(
    prefix: String,
    reason: String,
    runtime: SidecarRuntimeConfig?,
    startupTimeline: DesktopStartupTimeline? = nil,
    details: [String: Any] = [:]
  ) -> URL? {
    do {
      let directory = logsDirectory()
      try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
      let safePrefix = sanitizedFilePrefix(prefix)
      let fileURL = directory.appendingPathComponent("\(safePrefix)-\(timestampString()).json")
      let text = try make(
        runtime: runtime,
        reason: reason,
        startupTimeline: startupTimeline,
        details: details
      )
      try text.write(to: fileURL, atomically: true, encoding: .utf8)
      return fileURL
    } catch {
      NSLog("[Nexus Diagnostics] failed to write runtime diagnostics: \(error.localizedDescription)")
      return nil
    }
  }

  static func logsDirectory() -> URL {
    DesktopPaths.logsDirectory
  }

  private static func appPayload() -> [String: Any] {
    let bundle = Bundle.main
    return [
      "name": bundle.object(forInfoDictionaryKey: "CFBundleName") as? String ?? "Nexus",
      "bundle_identifier": bundle.bundleIdentifier ?? "",
      "version": bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0",
      "build_number": bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "dev",
      "bundle_path": bundle.bundleURL.path,
      "resource_path": bundle.resourceURL?.path ?? "",
      "executable_path": bundle.executableURL?.path ?? "",
      "url_schemes": urlSchemesPayload(bundle: bundle),
    ]
  }

  private static func processPayload() -> [String: Any] {
    let info = ProcessInfo.processInfo
    return [
      "pid": info.processIdentifier,
      "process_name": info.processName,
      "host_name": info.hostName,
      "current_directory": FileManager.default.currentDirectoryPath,
    ]
  }

  private static func systemPayload() -> [String: Any] {
    let version = ProcessInfo.processInfo.operatingSystemVersion
    return [
      "platform": "macos",
      "os_version": "\(version.majorVersion).\(version.minorVersion).\(version.patchVersion)",
      "active_processor_count": ProcessInfo.processInfo.activeProcessorCount,
      "physical_memory": ProcessInfo.processInfo.physicalMemory,
    ]
  }

  private static func runtimePayload(_ runtime: SidecarRuntimeConfig) -> [String: Any] {
    [
      "app_mode": runtime.appMode,
      "app_version": runtime.appVersion,
      "build_number": runtime.buildNumber,
      "platform": runtime.platform,
      "web_url": runtime.webURL.absoluteString,
      "api_base_url": runtime.apiBaseURL.absoluteString,
      "websocket_url": runtime.webSocketURL.absoluteString,
      "health_url": runtime.healthURL.absoluteString,
    ]
  }

  private static func pathsPayload() -> [String: Any] {
    return [
      "root_dir": DesktopPaths.rootDirectory.path,
      "data_dir": DesktopPaths.dataDirectory.path,
      "logs_dir": logsDirectory().path,
      "config_dir": DesktopPaths.configDirectory.path,
      "workspace_dir": DesktopPaths.workspaceDirectory.path,
      "cache_dir": DesktopPaths.cacheDirectory.path,
      "sidecar_pid_record": DesktopPaths.sidecarPIDFileURL.path,
      "connector_credentials_fallback_key": DesktopPaths.connectorCredentialsFallbackKeyURL.path,
    ]
  }

  private static func checksPayload() -> [String: Any] {
    let fileManager = FileManager.default
    return [
      "root_dir_exists": fileManager.fileExists(atPath: DesktopPaths.rootDirectory.path),
      "data_dir_exists": fileManager.fileExists(atPath: DesktopPaths.dataDirectory.path),
      "logs_dir_exists": fileManager.fileExists(atPath: logsDirectory().path),
      "sidecar_pid_record_exists": fileManager.fileExists(
        atPath: DesktopPaths.sidecarPIDFileURL.path
      ),
      "connector_credentials_fallback_key_exists": fileManager.fileExists(
        atPath: DesktopPaths.connectorCredentialsFallbackKeyURL.path
      ),
      "bundled_web_index_exists": bundledResourceExists(relativePath: "Web/index.html"),
      "bundled_sidecar_exists": bundledExecutableExists(name: "nexus-server"),
      "nexus_url_scheme_declared": declaredURLSchemes(bundle: Bundle.main).contains("nexus"),
    ]
  }

  private static func bundledResourceExists(relativePath: String) -> Bool {
    guard let resourceURL = Bundle.main.resourceURL else {
      return false
    }
    return FileManager.default.fileExists(atPath: resourceURL.appendingPathComponent(relativePath).path)
  }

  private static func bundledExecutableExists(name: String) -> Bool {
    guard let executableDir = Bundle.main.executableURL?.deletingLastPathComponent() else {
      return false
    }
    return FileManager.default.isExecutableFile(atPath: executableDir.appendingPathComponent(name).path)
  }

  private static func timestampString() -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyyMMdd-HHmmss"
    return formatter.string(from: Date())
  }

  private static func urlSchemesPayload(bundle: Bundle) -> [String] {
    declaredURLSchemes(bundle: bundle)
  }

  private static func declaredURLSchemes(bundle: Bundle) -> [String] {
    guard let urlTypes = bundle.object(forInfoDictionaryKey: "CFBundleURLTypes") as? [[String: Any]] else {
      return []
    }
    return urlTypes
      .flatMap { $0["CFBundleURLSchemes"] as? [String] ?? [] }
      .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
      .filter { !$0.isEmpty }
      .sorted()
  }

  private static func sanitizedFilePrefix(_ prefix: String) -> String {
    let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_")
    let value = prefix
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .unicodeScalars
      .map { allowed.contains($0) ? Character($0) : "-" }
    let text = String(value).trimmingCharacters(in: CharacterSet(charactersIn: "-_"))
    return text.isEmpty ? "runtime-issue" : text
  }

  private static func jsonString(_ value: Any) throws -> String {
    let data = try JSONSerialization.data(withJSONObject: value, options: [.prettyPrinted, .sortedKeys])
    guard let text = String(data: data, encoding: .utf8) else {
      throw DesktopDiagnosticsError.invalidJSON
    }
    return text
  }
}

private enum DesktopDiagnosticsError: LocalizedError {
  case invalidJSON

  var errorDescription: String? {
    switch self {
    case .invalidJSON:
      return "桌面诊断报告生成失败。"
    }
  }
}

File: desktop/windows/Nexus.Desktop/Diagnostics/DesktopDiagnosticsReport.cs
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;
using System.Text.Json;
using Microsoft.Win32;
using Nexus.Desktop.Runtime;
using Nexus.Desktop.Sidecar;

namespace Nexus.Desktop.Diagnostics;

internal static class DesktopDiagnosticsReport
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
    };

    public static string Make(
        SidecarRuntimeConfig? runtime,
        string? reason = null,
        DesktopStartupTimeline? startupTimeline = null,
        IReadOnlyDictionary<string, object?>? details = null)
    {
        var payload = new Dictionary<string, object?>
        {
            ["generated_at"] = DateTimeOffset.UtcNow.ToString("O"),
            ["app"] = AppPayload(),
            ["process"] = ProcessPayload(),
            ["system"] = SystemPayload(),
            ["paths"] = PathsPayload(),
            ["checks"] = ChecksPayload(),
        };

        if (runtime is not null)
        {
            payload["runtime"] = RuntimePayload(runtime);
        }
        if (!string.IsNullOrWhiteSpace(reason))
        {
            payload["reason"] = reason;
        }
        if (startupTimeline is not null)
        {
            payload["startup_timeline"] = startupTimeline.Snapshot().Select(TimelinePayload).ToList();
        }
        if (details is not null && details.Count > 0)
        {
            payload["details"] = details;
        }

        return JsonSerializer.Serialize(payload, JsonOptions);
    }

    public static string? WriteStartupFailure(Exception exception, DesktopStartupTimeline? startupTimeline = null)
    {
        return WriteReport(
            prefix: "startup-failure",
            reason: exception.Message,
            runtime: null,
            startupTimeline: startupTimeline,
            details: new Dictionary<string, object?>
            {
                ["exception_type"] = exception.GetType().FullName,
            });
    }

    public static string? WriteRuntimeIssue(
        string prefix,
        string reason,
        SidecarRuntimeConfig? runtime,
        DesktopStartupTimeline? startupTimeline = null,
        IReadOnlyDictionary<string, object?>? details = null)
    {
        return WriteReport(prefix, reason, runtime, startupTimeline, details);
    }

    private static string? WriteReport(
        string prefix,
        string reason,
        SidecarRuntimeConfig? runtime,
        DesktopStartupTimeline? startupTimeline,
        IReadOnlyDictionary<string, object?>? details)
    {
        try
        {
            Directory.CreateDirectory(DesktopPaths.LogsDirectory);
            string filePath = Path.Combine(DesktopPaths.LogsDirectory, $"{SanitizeFilePrefix(prefix)}-{DateTimeOffset.Now:yyyyMMdd-HHmmss}.json");
            string text = Make(runtime, reason, startupTimeline, details);
            File.WriteAllText(filePath, text);
            return filePath;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException or JsonException)
        {
            Trace.WriteLine($"[Nexus Diagnostics] failed to write diagnostics report: {exception.Message}");
            return null;
        }
    }

    private static Dictionary<string, object?> AppPayload()
    {
        return new Dictionary<string, object?>
        {
            ["name"] = "Nexus",
            ["version"] = AppVersionInfo.Version,
            ["build_number"] = AppVersionInfo.BuildNumber,
            ["base_directory"] = AppContext.BaseDirectory,
            ["executable_path"] = Environment.ProcessPath ?? string.Empty,
        };
    }

    private static Dictionary<string, object?> ProcessPayload()
    {
        using Process process = Process.GetCurrentProcess();
        return new Dictionary<string, object?>
        {
            ["pid"] = process.Id,
            ["process_name"] = process.ProcessName,
            ["current_directory"] = Environment.CurrentDirectory,
            ["user_interactive"] = Environment.UserInteractive,
        };
    }

    private static Dictionary<string, object?> SystemPayload()
    {
        return new Dictionary<string, object?>
        {
            ["platform"] = "windows",
            ["os_version"] = RuntimeInformation.OSDescription,
            ["architecture"] = RuntimeInformation.OSArchitecture.ToString(),
            ["process_architecture"] = RuntimeInformation.ProcessArchitecture.ToString(),
            ["machine_name"] = Environment.MachineName,
            ["processor_count"] = Environment.ProcessorCount,
            ["is_64bit_process"] = Environment.Is64BitProcess,
        };
    }

    private static Dictionary<string, object?> RuntimePayload(SidecarRuntimeConfig runtime)
    {
        return new Dictionary<string, object?>
        {
            ["app_mode"] = runtime.AppMode,
            ["app_version"] = runtime.AppVersion,
            ["build_number"] = runtime.BuildNumber,
            ["platform"] = runtime.Platform,
            ["web_url"] = runtime.WebBaseUrl,
            ["api_base_url"] = runtime.ApiBaseUrl,
            ["websocket_url"] = runtime.WebSocketUrl,
            ["health_url"] = runtime.HealthUrl,
        };
    }

    private static Dictionary<string, object?> PathsPayload()
    {
        return new Dictionary<string, object?>
        {
            ["application_data_dir"] = DesktopPaths.ApplicationDataDirectory,
            ["config_dir"] = DesktopPaths.ConfigDirectory,
            ["workspace_dir"] = DesktopPaths.WorkspaceDirectory,
            ["cache_dir"] = DesktopPaths.CacheDirectory,
            ["logs_dir"] = DesktopPaths.LogsDirectory,
            ["connector_credentials_dpapi"] = Path.Combine(DesktopPaths.ConfigDirectory, "connector-credentials.dpapi"),
            ["connector_credentials_fallback_key"] = Path.Combine(DesktopPaths.ConfigDirectory, "connector-credentials.key"),
        };
    }

    private static Dictionary<string, object?> ChecksPayload()
    {
        string resourceDirectory = Path.Combine(AppContext.BaseDirectory, "Resources");
        return new Dictionary<string, object?>
        {
            ["application_data_exists"] = Directory.Exists(DesktopPaths.ApplicationDataDirectory),
            ["logs_dir_exists"] = Directory.Exists(DesktopPaths.LogsDirectory),
            ["bundled_web_app_exists"] = File.Exists(Path.Combine(resourceDirectory, "Web", "app.html")),
            ["bundled_web_settings_exists"] = File.Exists(Path.Combine(resourceDirectory, "Web", "settings.html")),
            ["bundled_web_oauth_callback_exists"] = File.Exists(Path.Combine(resourceDirectory, "Web", "oauth-callback.html")),
            ["bundled_sidecar_exists"] = File.Exists(Path.Combine(resourceDirectory, "nexus-server.exe")),
            ["connector_credentials_dpapi_exists"] = File.Exists(Path.Combine(DesktopPaths.ConfigDirectory, "connector-credentials.dpapi")),
            ["connector_credentials_fallback_key_exists"] = File.Exists(Path.Combine(DesktopPaths.ConfigDirectory, "connector-credentials.key")),
            ["nexus_url_scheme_registered"] = NexusUrlSchemeRegistered(),
        };
    }

    private static Dictionary<string, object?> TimelinePayload(DesktopStartupEvent item)
    {
        return new Dictionary<string, object?>
        {
            ["name"] = item.Name,
            ["elapsed_ms"] = item.ElapsedMilliseconds,
            ["delta_ms"] = item.DeltaMilliseconds,
            ["metadata"] = item.Metadata,
        };
    }

    private static bool NexusUrlSchemeRegistered()
    {
        try
        {
            using RegistryKey? key = Registry.CurrentUser.OpenSubKey(@"Software\Classes\nexus");
            return key?.GetValue("URL Protocol") is not null;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            return false;
        }
    }

    private static string SanitizeFilePrefix(string prefix)
    {
        char[] chars = prefix
            .Trim()
            .Select(item => char.IsLetterOrDigit(item) || item is '-' or '_' ? item : '-')
            .ToArray();
        string value = new string(chars).Trim('-', '_');
        return string.IsNullOrWhiteSpace(value) ? "runtime-issue" : value;
    }
}
